compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1418 @@
+    val owner = klass
+
+    def show(using Context) = "ObjectRef(" + klass.show + ")"
+
+  /**
+   * Represents values that are instances of the specified class.
+   *
+   * Note that the 2nd parameter block does not take part in the definition of equality.
+   */
+  case class OfClass private (
+    klass: ClassSymbol, outer: Value, ctor: Symbol, args: List[Value], env: Env.Data)(
+    valsMap: mutable.Map[Symbol, Value], varsMap: mutable.Map[Symbol, Heap.Addr], outersMap: mutable.Map[ClassSymbol, Value])
+  extends Ref(valsMap, varsMap, outersMap):
+    def widenedCopy(outer: Value, args: List[Value], env: Env.Data): OfClass =
+      new OfClass(klass, outer, ctor, args, env)(this.valsMap, this.varsMap, this.outersMap)
+
+    def show(using Context) =
+      val valFields = vals.map(_.show +  " -> " +  _.show)
+      "OfClass(" + klass.show + ", outer = " + outer + ", args = " + args.map(_.show) + ", vals = " + valFields + ")"
+
+  object OfClass:
+    def apply(
+      klass: ClassSymbol, outer: Value, ctor: Symbol, args: List[Value], env: Env.Data)(
+      using Context
+    ): OfClass =
+      val instance = new OfClass(klass, outer, ctor, args, env)(
+        valsMap = mutable.Map.empty, varsMap = mutable.Map.empty, outersMap = mutable.Map.empty
+      )
+      instance.initOuter(klass, outer)
+      instance
+
+  /**
+   * Represents arrays.
+   *
+   * Note that the 2nd parameter block does not take part in the definition of equality.
+   *
+   * Different arrays are distinguished by the context. Currently the default context is the static
+   * object whose initialization triggers the creation of the array.
+   *
+   * In the future, it is possible that we introduce a mechanism for end-users to mark the context.
+   *
+   * @param owner The static object whose initialization creates the array.
+   */
+  case class OfArray(owner: ClassSymbol, regions: Regions.Data)(using @constructorOnly ctx: Context)
+  extends Ref(valsMap = mutable.Map.empty, varsMap = mutable.Map.empty, outersMap = mutable.Map.empty):
+    val klass: ClassSymbol = defn.ArrayClass
+    val addr: Heap.Addr = Heap.arrayAddr(regions, owner)
+    def show(using Context) = "OfArray(owner = " + owner.show + ")"
+
+  /**
+   * Represents a lambda expression
+   */
+  case class Fun(code: Tree, thisV: Value, klass: ClassSymbol, env: Env.Data) extends Value:
+    def show(using Context) = "Fun(" + code.show + ", " + thisV.show + ", " + klass.show + ")"
+
+  /**
+   * Represents a set of values
+   *
+   * It comes from `if` expressions.
+   */
+  case class RefSet(refs: ListSet[Value]) extends Value:
+    assert(refs.forall(!_.isInstanceOf[RefSet]))
+    def show(using Context) = refs.map(_.show).mkString("[", ",", "]")
+
+  /** A cold alias which should not be used during initialization. */
+  case object Cold extends Value:
+    def show(using Context) = "Cold"
+
+  val Bottom = RefSet(ListSet.empty)
```

Comment:
@liufengyun Why do we set `Bottom` as an empty `RefSet`? Is it because it can be easily joined with other `RefSet`?

Reply:
We use it as a convenience so that we have fewer cases in the method `call/select/assign`. The convenience should work as if we define a separate `Bottom` --- otherwise, it's broken.